aspnetcore/web-api/define-controller.md

```diff
@@ -0,0 +1,113 @@
+
+[!code-csharp[](../web-api/define-controller/samples/WebApiSample.Api.Pre21/Controllers/OrdersController.cs?name=snippet_OrdersController&highlight=1)]
+
+In the preceding controller, the `Index` action returns a [ViewResult](/dotnet/api/microsoft.aspnetcore.mvc.viewresult) representing the associated MVC view at *Views/Orders/Index.cshtml*. The `GetById` and `CreateAsync` actions respond to HTTP GET and POST requests, respectively.
+
+## Derive class from ControllerBase
+
+Inherit from the [ControllerBase](/dotnet/api/microsoft.aspnetcore.mvc.controllerbase) class when your controller doesn't need to return MVC views or [invoke view components]((xref:mvc/views/view-components#invoking-a-view-component-directly-from-a-controller)). For example, the following controller only supports Web API actions:
+
+[!code-csharp[](../web-api/define-controller/samples/WebApiSample.Api.Pre21/Controllers/PetsController.cs?name=snippet_PetsController&highlight=3)]
+
+A benefit of deriving from `ControllerBase` instead of `Controller` is that IntelliSense displays only Web API-specific members.
+
+::: moniker range=">= aspnetcore-2.1"
+## Annotate class with ApiControllerAttribute
+
+ASP.NET Core 2.1 introduces the `[ApiController]` attribute to denote a Web API controller class. For example:
+
+[!code-csharp[](../web-api/define-controller/samples/WebApiSample.Api/Controllers/ProductsController.cs?name=snippet_ControllerSignature&highlight=2)]
+
+This attribute is commonly coupled with either `ControllerBase` or `Controller` to gain access to useful methods and properties. `ControllerBase` provides access to methods such as [CreatedAtAction](/dotnet/api/microsoft.aspnetcore.mvc.controllerbase.createdataction) and [File](/dotnet/api/microsoft.aspnetcore.mvc.controllerbase.file). `Controller` provides access to methods such as [Json](/dotnet/api/microsoft.aspnetcore.mvc.controller.json) and [View](/dotnet/api/microsoft.aspnetcore.mvc.controller.view).
+
+The following sections describe convenience features added by the attribute.
+
+### Automatic HTTP 400 responses
+
+Validation errors automatically trigger an HTTP 400 response. The following code becomes unnecessary in your actions:
+
+[!code-csharp[](../web-api/define-controller/samples/WebApiSample.Api.Pre21/Controllers/PetsController.cs?range=46-49)]
+
+This default behavior is disabled with the following code in *Startup.ConfigureServices*:
+
+[!code-csharp[](../web-api/define-controller/samples/WebApiSample.Api/Startup.cs?name=snippet_ConfigureApiBehaviorOptions&highlight=5)]
+
+### Binding source parameter inference
+
+Inference rules are applied for the default data sources of action parameters. The binding source attributes behave as follows:
```

Comment:
I think before digging into the details it would be good to show an example and explain the common patterns. Most APIs will bind a parameter (or two from the route) - usually an int or string, and a parameter or two from the query string - usually a simple type, and usually a  complex type from the body.

It would be good to show how the `[ApiController]` attribute cleans up the code by inferring the details of the idiomatic approach for binding in APIs